File: backend/src/AgentTeam/Models/WorkflowSchema.php
<?php

declare(strict_types=1);

namespace AgentTeam\Models;

/**
 * WorkflowSchema Model
 *
 * Represents a reusable JSON Schema definition that can be attached to
 * agent workflow nodes for constrained decoding (structured outputs).
 *
 * Stored in the workflow_schemas table.
 */
class WorkflowSchema
{
    private ?int $id = null;
    private int $userId = 0;
    private string $name = '';
    private string $description = '';
    private array $schemaJson = [];
    private bool $strict = true;
    private ?string $createdAt = null;
    private ?string $updatedAt = null;

    public function __construct(array $data = [])
    {
        if (!empty($data)) {
            $this->hydrate($data);
        }
    }

    public function hydrate(array $data): self
    {
        $this->id = isset($data['id']) ? (int) $data['id'] : null;
        $this->userId = isset($data['user_id']) ? (int) $data['user_id'] : 0;
        $this->name = $data['name'] ?? '';
        $this->description = $data['description'] ?? '';
        $this->schemaJson = $this->decodeJson($data['schema_json'] ?? []);
        $this->strict = (bool) ($data['strict'] ?? true);
        $this->createdAt = $data['created_at'] ?? null;
        $this->updatedAt = $data['updated_at'] ?? null;

        return $this;
    }

    private function decodeJson($value): array
    {
        if (is_string($value) && $value !== '') {
            $decoded = json_decode($value, true);
            return is_array($decoded) ? $decoded : [];
        }
        return is_array($value) ? $value : [];
    }

    public function toArray(): array
    {
        return [
            'id' => $this->id,
            'user_id' => $this->userId,
            'name' => $this->name,
            'description' => $this->description,
            'schema_json' => $this->schemaJson,
            'strict' => $this->strict,
            'created_at' => $this->createdAt,
            'updated_at' => $this->updatedAt,
        ];
    }

    public function toApiArray(): array
    {
        return [
            'id' => $this->id,
            'name' => $this->name,
            'description' => $this->description,
            'schema_json' => $this->schemaJson,
            'strict' => $this->strict,
            'created_at' => $this->createdAt,
            'updated_at' => $this->updatedAt,
        ];
    }

    /**
     * Build the canonical "output_schema" structure that gets passed
     * to providers for constrained decoding.
     */
    public function toOutputSchema(): array
    {
        return [
            'name' => $this->name,
            'description' => $this->description,
            'strict' => $this->strict,
            'schema' => $this->schemaJson,
        ];
    }

    /**
     * Validate that schema_json is a non-empty JSON Schema-like structure.
     * Returns array of error messages (empty = valid).
     */
    public function validate(): array
    {
        $errors = [];

        if ($this->name === '') {
            $errors[] = 'Schema name is required';
        }
        if (!preg_match('/^[a-zA-Z0-9_\-]+$/', $this->name)) {
            $errors[] = 'Schema name must contain only letters, numbers, dashes and underscores';
        }
        if (empty($this->schemaJson)) {
            $errors[] = 'schema_json is required';
        } else {
            $type = $this->schemaJson['type'] ?? null;
            if ($type !== 'object') {
                $errors[] = 'Top-level schema type must be "object"';
            }
            if (!isset($this->schemaJson['properties']) || !is_array($this->schemaJson['properties'])) {
                $errors[] = 'Schema must have a "properties" object';
            }
        }

        return $errors;
    }

    // Getters
    public function getId(): ?int { return $this->id; }
    public function getUserId(): int { return $this->userId; }
    public function getName(): string { return $this->name; }
    public function getDescription(): string { return $this->description; }
    public function getSchemaJson(): array { return $this->schemaJson; }
    public function isStrict(): bool { return $this->strict; }

    // Setters
    public function setId(?int $id): self { $this->id = $id; return $this; }
    public function setUserId(int $userId): self { $this->userId = $userId; return $this; }
    public function setName(string $name): self { $this->name = $name; return $this; }
    public function setDescription(string $description): self { $this->description = $description; return $this; }
    public function setSchemaJson(array $schemaJson): self { $this->schemaJson = $schemaJson; return $this; }
    public function setStrict(bool $strict): self { $this->strict = $strict; return $this; }
}

File: backend_python/app/agent_team/models/workflow_schema.py
"""Port of backend/src/AgentTeam/Models/WorkflowSchema.php.

Represents a reusable JSON Schema definition that can be attached to
agent workflow nodes for constrained decoding (structured outputs).

Stored in the workflow_schemas table.
"""
from __future__ import annotations

import json
import re

from app.support.phpcompat import php_array, php_bool, php_empty, php_intval

_NAME_RE = re.compile(r'^[a-zA-Z0-9_\-]+$')


def _decode_json(value):
    """Mirror WorkflowSchema::decodeJson (PHP 47-54)."""
    if isinstance(value, str) and value != '':
        try:
            decoded = json.loads(value)
        except ValueError:
            decoded = None
        return decoded if isinstance(decoded, (list, dict)) else []
    return value if isinstance(value, (list, dict)) else []


class WorkflowSchema:
    def __init__(self, data: dict | None = None):
        """PHP: `if (!empty($data)) { $this->hydrate($data); }`."""
        self.id: int | None = None
        self.userId: int = 0
        self.name: str = ''
        self.description: str = ''
        self.schemaJson: list = []
        self.strict: bool = True
        self.createdAt: str | None = None
        self.updatedAt: str | None = None

        if data:
            self.hydrate(data)

    def hydrate(self, data: dict) -> 'WorkflowSchema':
        self.id = php_intval(data['id']) if data.get('id') is not None else None
        self.userId = php_intval(data['user_id']) if data.get('user_id') is not None else 0
        self.name = data.get('name') if data.get('name') is not None else ''
        self.description = data.get('description') if data.get('description') is not None else ''
        schema_json = data.get('schema_json')
        self.schemaJson = _decode_json(schema_json if schema_json is not None else [])
        strict = data.get('strict')
        self.strict = php_bool(strict if strict is not None else True)
        self.createdAt = data.get('created_at') if data.get('created_at') is not None else None
        self.updatedAt = data.get('updated_at') if data.get('updated_at') is not None else None

        return self

    def toArray(self) -> dict:
        return {
            'id': self.id,
            'user_id': self.userId,
            'name': self.name,
            'description': self.description,
            'schema_json': php_array(self.schemaJson),
            'strict': self.strict,
            'created_at': self.createdAt,
            'updated_at': self.updatedAt,
        }

    def toApiArray(self) -> dict:
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'schema_json': php_array(self.schemaJson),
            'strict': self.strict,
            'created_at': self.createdAt,
            'updated_at': self.updatedAt,
        }

    def toOutputSchema(self) -> dict:
        """Build the canonical "output_schema" structure that gets passed
        to providers for constrained decoding."""
        return {
            'name': self.name,
            'description': self.description,
            'strict': self.strict,
            'schema': php_array(self.schemaJson),
        }

    def validate(self) -> list:
        """Validate that schema_json is a non-empty JSON Schema-like structure.
        Returns list of error messages (empty = valid)."""
        errors = []

        if self.name == '':
            errors.append('Schema name is required')
        if not _NAME_RE.match(self.name):
            errors.append('Schema name must contain only letters, numbers, dashes and underscores')
        if php_empty(self.schemaJson):
            errors.append('schema_json is required')
        else:
            schema = self.schemaJson if isinstance(self.schemaJson, dict) else {}
            schema_type = schema.get('type')
            if schema_type != 'object':
                errors.append('Top-level schema type must be "object"')
            properties = schema.get('properties')
            if properties is None or not isinstance(properties, (list, dict)):
                errors.append('Schema must have a "properties" object')

        return errors

    # Getters
    def getId(self) -> int | None:
        return self.id

    def getUserId(self) -> int:
        return self.userId

    def getName(self) -> str:
        return self.name

    def getDescription(self) -> str:
        return self.description

    def getSchemaJson(self) -> list:
        return self.schemaJson

    def isStrict(self) -> bool:
        return self.strict

    # Setters
    def setId(self, id: int | None) -> 'WorkflowSchema':
        self.id = id
        return self

    def setUserId(self, user_id: int) -> 'WorkflowSchema':
        self.userId = user_id
        return self

    def setName(self, name: str) -> 'WorkflowSchema':
        self.name = name
        return self

    def setDescription(self, description: str) -> 'WorkflowSchema':
        self.description = description
        return self

    def setSchemaJson(self, schema_json: dict) -> 'WorkflowSchema':
        self.schemaJson = schema_json
        return self

    def setStrict(self, strict: bool) -> 'WorkflowSchema':
        self.strict = strict
        return self
